src/GlassCat/GlassCat.jl

```diff
@@ -20,7 +20,11 @@ include("Air.jl")
 export Air, isair
 
 # include built glass cat source files
-include("data/jl/AGFGlassCat.jl")
+@assert AGFGLASSCAT_PATH === joinpath(@__DIR__, "data", "jl", "AGFGlassCat.jl")
+if !isfile(AGFGLASSCAT_PATH)
+    @error "$(basename(AGFGLASSCAT_PATH)) not found! Run `] build` to trigger the build steps."
```

Comment:
If we use `add_agf` at runtime it must rebuild the catalog for the glasses to be usable - doesn't this mean we can just rebuild at runtime here automatically in the same way, rather than making the user do it manually?

Reply:
Yeah I suppose we can. I said it was messy in my reply to @BrianGun because this will re-`include` a few GlassCat source files from inside the GlassCat module, which re-defines some constants. But it's probably alright.

Reply:
These are the kinds of warnings I got when I tried it out locally.

```
julia> using OpticSim
[ Info: Precompiling OpticSim [24114763-4efb-45e7-af0e-cde916beb153]
[ Info: AGFGlassCat.jl not found! Running build steps...
WARNING: redefinition of constant DISPFORM_NAMES. This may fail, cause incorrect answers, or produce other errors.
WARNING: redefinition of constant STATUS. This may fail, cause incorrect answers, or produce other errors.
WARNING: Method definition (::Type{OpticSim.GlassCat.GlassType})(Integer) in module GlassCat at Enums.jl:191 overwritten on the same line (check for 
duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition namemap(Type{OpticSim.GlassCat.GlassType}) in module GlassCat at Enums.jl:195 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition typemin(Type{OpticSim.GlassCat.GlassType}) in module GlassCat at Enums.jl:196 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition typemax(Type{OpticSim.GlassCat.GlassType}) in module GlassCat at Enums.jl:197 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition instances(Type{OpticSim.GlassCat.GlassType}) in module GlassCat at Enums.jl:199 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition (::Type{OpticSim.GlassCat.GlassID})(OpticSim.GlassCat.GlassType, Int64) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:14 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition (::Type{OpticSim.GlassCat.GlassID})(Any, Any) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:14 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

┌ Warning: Replacing docs for `OpticSim.GlassCat.GlassID :: Union{}` in module `OpticSim.GlassCat`
└ @ Base.Docs docs\Docs.jl:240
WARNING: Method definition show(IO, OpticSim.GlassCat.GlassID) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:18 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition type(OpticSim.GlassCat.GlassID) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:19 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition num(OpticSim.GlassCat.GlassID) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:20 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition ismodel(OpticSim.GlassCat.GlassID) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:22 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition isMIL(OpticSim.GlassCat.GlassID) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:23 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition isAGF(OpticSim.GlassCat.GlassID) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:24 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition isother(OpticSim.GlassCat.GlassID) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:25 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

┌ Warning: Replacing docs for `OpticSim.GlassCat.AbstractGlass :: Union{}` in module `OpticSim.GlassCat`
└ @ Base.Docs docs\Docs.jl:240
WARNING: Method definition (::Type{OpticSim.GlassCat.Glass})(OpticSim.GlassCat.GlassID, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:79 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

┌ Warning: Replacing docs for `OpticSim.GlassCat.Glass :: Union{}` in module `OpticSim.GlassCat`
└ @ Base.Docs docs\Docs.jl:240
WARNING: Method definition glassid(OpticSim.GlassCat.Glass) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:98 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

┌ Warning: Replacing docs for `OpticSim.GlassCat.glassid :: Tuple{OpticSim.GlassCat.Glass}` in module `OpticSim.GlassCat`
└ @ Base.Docs docs\Docs.jl:240
WARNING: Method definition glassname(OpticSim.GlassCat.Glass) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:105 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

┌ Warning: Replacing docs for `OpticSim.GlassCat.glassname :: Tuple{OpticSim.GlassCat.Glass}` in module `OpticSim.GlassCat`
└ @ Base.Docs docs\Docs.jl:240
WARNING: Method definition glassname(OpticSim.GlassCat.GlassID) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:106 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition show(IO, OpticSim.GlassCat.Glass) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:120 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition glassforid(OpticSim.GlassCat.GlassID) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:129 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

┌ Warning: Replacing docs for `OpticSim.GlassCat.glassforid :: Tuple{OpticSim.GlassCat.GlassID}` in module `OpticSim.GlassCat`
└ @ Base.Docs docs\Docs.jl:240
WARNING: Method definition info(IO, OpticSim.GlassCat.Glass) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:181 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

┌ Warning: Replacing docs for `OpticSim.GlassCat.info :: Tuple{IO, OpticSim.GlassCat.Glass}` in module `OpticSim.GlassCat`
└ @ Base.Docs docs\Docs.jl:240
WARNING: Method definition info(OpticSim.GlassCat.AbstractGlass) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:337 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition docstring(IO, OpticSim.GlassCat.Glass) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\GlassTypes.jl:339 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

[ Info: [✓] Verified file at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\data\agf\NIKON.agf
[ Info: [✓] Verified file at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\data\agf\OHARA.agf
[ Info: [✓] Verified file at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\data\agf\HOYA.agf
[ Info: [✓] Verified file at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\data\agf\SCHOTT.agf
[ Info: [✓] Verified file at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\data\agf\Sumita.agf
[ Info: Using sources: NIKON, OHARA, HOYA, SCHOTT and Sumita
[ Info: Writing NIKON.jl
[ Info: Writing OHARA.jl
[ Info: Writing HOYA.jl
[ Info: Writing SCHOTT.jl
[ Info: Writing Sumita.jl
[ Info: Writing AGFGlassCat.jl
WARNING: Method definition add_agf(AbstractString) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\sources.jl:18 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition add_agf##kw(Any, typeof(OpticSim.GlassCat.add_agf), AbstractString) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\sources.jl:18 overwritten on the same line (check for duplicate calls to `include`).

┌ Warning: Replacing docs for `OpticSim.GlassCat.add_agf :: Tuple{AbstractString}` in module `OpticSim.GlassCat`
└ @ Base.Docs docs\Docs.jl:240
WARNING: Method definition verify_sources!(AbstractArray{var"#s32", 1} where var"#s32"<:(AbstractArray{var"#s31", 1} where var"#s31"<:AbstractString), AbstractString) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\sources.jl:63 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

┌ Warning: Replacing docs for `OpticSim.GlassCat.verify_sources! :: Tuple{AbstractVector{var"#s27"} where var"#s27"<:(AbstractVector{var"#s26"} where var"#s26"<:AbstractString), AbstractString}` in module `OpticSim.GlassCat`
└ @ Base.Docs docs\Docs.jl:240
WARNING: Method definition verify_source(AbstractString, AbstractString) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\sources.jl:89 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

┌ Warning: Replacing docs for `OpticSim.GlassCat.verify_source :: Tuple{AbstractString, AbstractString}` in module `OpticSim.GlassCat`
└ @ Base.Docs docs\Docs.jl:240
WARNING: Method definition download_source(AbstractString, AbstractString) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\sources.jl:107 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition download_source(AbstractString, AbstractString, Union{Nothing, AbstractString}) in module GlassCat at C:\Users\v-alfredwong\Documents\OpticSim.jl\src\GlassCat\sources.jl:107 overwritten on the same line (check for duplicate calls to `include`).
  ** incremental compilation may be fatally broken for this module **

┌ Warning: Replacing docs for `OpticSim.GlassCat.download_source :: Union{Tuple{AbstractString, AbstractString}, Tuple{AbstractString, AbstractString, Union{Nothing, AbstractString}}}` in module `OpticSim.GlassCat`
└ @ Base.Docs docs\Docs.jl:240
```

Reply:
Hmm a good reason to avoid it - if the file doesn't exist to start with though why would things be redefined?
I guess the answer is probably to follow Julia precedent here - is it typical for packages to error out telling the user to run `] build my_pkg`? I think I've had this before, in which case we should follow that and do as you have done here rather than automatically run the build.

Doesn't the redefinition problem exist for `add_agf` too or do we do something a bit smarter in that case?

Reply:
I think I remember seeing something like this in GR.jl, which is why I had this idea.

I still get the first 2 warning lines when using `add_agf`. I was going to address this, but the PR was merged a bit prematurely (in hindsight, I should've made a draft PR).

I think we see the longer warning list in this case because the build steps are being executed in the global scope. Not entirely sure. It may be best to remove the automatic rebuild option in `add_agf` and also prompt the user to run a build.

Other options I can think of are (1) to create a temporary environment from which to run the build, which avoids redefinitions, or (2) to do `Pkg.build` instead of `include`. I'm pretty busy with Bruges today, so I'll try these out over the weekend.

Reply:
I think I'd expect the behavior of `add_agf` to make the added stuff available immediately (without having to explicitly build). Both these options sound reasonable - I think it should a priority to make it graceful for `add_agf`. If the solution also works for the first `using` then that's great, otherwise warning (as for GR) seems like a reasonable option in that case.

Reply:
Automatic rebuilding/reloading is ideal from a usability standpoint. Seems like exactly the kind of thing Revise.jl was designed for.

4cbc74a should solve everything :)